File: Recursion/4. Min Cost Climbing Stairs/min_cost_climbing_stairs.cpp
// https://leetcode.com/problems/min-cost-climbing-stairs
// (TLE solutions - not the most optimised solutions)

//Solution 1

class Solution {
public:

    int helper(int ind, vector<int>& cost) {
        if(ind <= 1) return cost[ind];
        return ((ind == cost.size() ? 0 : cost[ind]) + min(helper(ind-1, cost), helper(ind-2, cost))); 
    }

    int minCostClimbingStairs(vector<int>& cost) {
        return helper(cost.size() , cost);
    }
};

//Solution 2

class Solution {
public:

    int helper(int ind, vector<int>& cost) {
        if(ind <= 1) return cost[ind];
        return (cost[ind] + min(helper(ind-1, cost), helper(ind-2, cost))); 
    }

    int minCostClimbingStairs(vector<int>& cost) {
        return min(helper(cost.size() - 1 , cost), helper(cost.size() - 2, cost));
    }
};

//Solution 3 - From 0 to n

class Solution {
public:
    int helper(vector<int>& cost, int ind) {
        if(ind >= cost.size()) return 0;
        return cost[ind] + min(helper(cost, ind+1), helper(cost, ind+2));
    }

    int minCostClimbingStairs(vector<int>& cost) {
        return min(helper(cost, 0), helper(cost, 1));
    }
};

File: Recursion/4. Min Cost Climbing Stairs/min_cost_climbing_stairs.py
# https://leetcode.com/problems/min-cost-climbing-stairs
# (TLE solutions - not the most optimised solutions)

# Bottom up approach
# Actual problem: Find min cost to reach top of stairs
# Sub problem: find min cost to reach n+1 or n+2 steps
# Recurrence relation: f(n)=min(cost[n]+f(n+1),cost[n+1]+f(n+2))
# Time Complexity: O(2^n)
# Space Complexity: O(n)

class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        return min(self.costToClimbStairs(0,cost),self.costToClimbStairs(1,cost))  # since we can start from either step 0 or step 1
    def costToClimbStairs(self,currentStep: int,cost: List[int])-> int:
        if currentStep>= len(cost):  # base condition
            return 0
        return cost[currentStep]+min(self.costToClimbStairs(currentStep+1,cost),self.costToClimbStairs(currentStep+2,cost))




# Top down approach
# Recurrence relation: f(n)=min(cost[n]+f(n-1),cost[n-1]+f(n-2))
class Solution:
    def minCostClimbingStairs(self, cost: List[int]) -> int:
        return min(self.costToClimbStairs(len(cost)-1,cost),self.costToClimbStairs(len(cost)-2,cost)) 
    def costToClimbStairs(self,currentStep: int,cost: List[int])-> int:
        if currentStep<=1:  # base condition
            return cost[currentStep]
        return cost[currentStep]+min(self.costToClimbStairs(currentStep-1,cost),self.costToClimbStairs(currentStep-2,cost))